core/src/main/java/com/graphhopper/routing/ch/CHRoutingAlgorithmFactory.java

```diff
@@ -0,0 +1,107 @@
+        this.chGraph = chGraph;
+        this.chProfile = chGraph.getCHProfile();
+        prepareWeighting = new PreparationWeighting(chProfile.getWeighting());
+    }
+
+    @Override
+    public RoutingAlgorithm createAlgo(Graph graph, AlgorithmOptions opts) {
+        // todo: This method does not really fit for CH: We are passed a graph, but really we already know which
+        // graph we have to use: the CH graph. Same with  opts.weighting: The CHProfile already contains a weighting
+        // and we cannot really use it here. The real reason we do this the way its done atm is that graph might be
+        // a QueryGraph that wraps (our) CHGraph.
+        AbstractBidirAlgo algo = doCreateAlgo(graph, opts);
+        algo.setEdgeFilter(new LevelEdgeFilter(chGraph));
+        algo.setMaxVisitedNodes(opts.getMaxVisitedNodes());
+        return algo;
+    }
+
+    private AbstractBidirAlgo doCreateAlgo(Graph graph, AlgorithmOptions opts) {
+        if (chProfile.isEdgeBased()) {
+            return createAlgoEdgeBased(graph, opts);
+        } else {
+            return createAlgoNodeBased(graph, opts);
+        }
+    }
+
+    private AbstractBidirAlgo createAlgoEdgeBased(Graph graph, AlgorithmOptions opts) {
+        if (ASTAR_BI.equals(opts.getAlgorithm())) {
+            return new AStarBidirectionEdgeCHNoSOD(graph, createTurnWeightingForEdgeBased(graph))
+                    .setApproximation(RoutingAlgorithmFactorySimple.getApproximation(ASTAR_BI, opts, graph.getNodeAccess()));
+        } else if (DIJKSTRA_BI.equals(opts.getAlgorithm())) {
+            return new DijkstraBidirectionEdgeCHNoSOD(graph, createTurnWeightingForEdgeBased(graph));
+        } else {
+            throw new IllegalArgumentException("Algorithm " + opts.getAlgorithm() + " not supported for edge-based Contraction Hierarchies. Try with ch.disable=true");
+        }
+    }
+
+    private AbstractBidirAlgo createAlgoNodeBased(Graph graph, AlgorithmOptions opts) {
+        if (ASTAR_BI.equals(opts.getAlgorithm())) {
+            return new AStarBidirectionCH(graph, prepareWeighting)
+                    .setApproximation(RoutingAlgorithmFactorySimple.getApproximation(ASTAR_BI, opts, graph.getNodeAccess()));
+        } else if (DIJKSTRA_BI.equals(opts.getAlgorithm())) {
+            if (opts.getHints().getBool("stall_on_demand", true)) {
+                return new DijkstraBidirectionCH(graph, prepareWeighting);
+            } else {
+                return new DijkstraBidirectionCHNoSOD(graph, prepareWeighting);
+            }
+        } else {
+            throw new IllegalArgumentException("Algorithm " + opts.getAlgorithm() + " not supported for node-based Contraction Hierarchies. Try with ch.disable=true");
+        }
+    }
+
+    private TurnWeighting createTurnWeightingForEdgeBased(Graph graph) {
+        // important: do not simply take the turn cost storage from ghStorage, because we need the wrapped storage from
```

Comment:
Do you know what this comment was meant for?

Reply:
Yes. If  `graph` is a `QueryGraph` (not just a `CHGraph`) we have to use `QueryGraphTurnCostStorage` which provides some special handling for virtual nodes (do not apply turn costs) and converts the virtual edge ids to the underlying real edge ids before getting the turn costs from the actual turn cost storage. This is a problem also for #1820, because it means that `TurnWeighting` is not just another `Weighting` but it uses information that is only available after the query graph has been constructed. This works because currently we always create a `TurnWeighting` for each request (compared to the `Weighting` objects that always live in `CHDecorator/CHProfile`), but if we want to include `calcTurnWeight` etc. in `Weighting` this is a problem.

Reply:
Now I understand - thanks!

> This is a problem also for #1820

Yeah, I stumbled over this already in #1775

> it means that TurnWeighting is not just another Weighting but it uses information that is only available after the query graph has been constructed

This is interesting. The BlockAreaWeighting is similar: it requires a query graph. But when I thought about it in more detail I found that it could use a base graph and just throw an exception if query points are inside the blocked areas, which should help to resolve the dependency problem (QueryGraph->BlockAreaWeighting, LocationIndex->BlockAreaWeighting and LocationIndex->QueryGraph).

> but if we want to include calcTurnWeight etc. in Weighting this is a problem.

You mean for CH this is a problem? Because for non-CH we already create a Weighting per request.

Reply:
> You mean for CH this is a problem? Because for non-CH we already create a Weighting per request.

Yes I meant for CH (or at least this is the one place where I know its problematic, not sure if there are other places).